src/traces/heatmap/clean_2d_array.js

```diff
@@ -30,12 +31,38 @@ module.exports = function clean2dArray(zOld, transpose) {
         old2new = function(zOld, i, j) { return zOld[i][j]; };
     }
 
+    var padOld2new = function(zOld, i, j) {
+        if(i === undefined || j === undefined) return undefined;
+        return old2new(zOld, i, j);
+    };
+
+    function axisMapping(ax) {
+        if(trace && trace.type !== 'carpet' && trace.type !== 'contourcarpet' &&
+            ax && ax.type === 'category' && trace['_' + ax._id.charAt(0)].length) {
+            var axMapping = [];
+            for(i = 0; i < ax._categories.length; i++) {
+                axMapping.push((trace['_' + ax._id.charAt(0) + 'Map'] || trace[ax._id.charAt(0)]).indexOf(ax._categories[i]));
+            }
+            return function(i) {return axMapping[i] === -1 ? undefined : axMapping[i];};
```

Comment:
You should be using `var BADNUM = require('../../constants/numerical').BADNUM` instead of `undefined` here.

Reply:
> Why is that a big deal? I don't see a problem with turning `typeof number` items in `_categories` into strings if that makes things easier.

The problem is that we need a mechanism to check whether a trace has data associated with a given category. That is why we do an `indexOf` call. Of course, if the category and the input data have a different type, this lookup will not work properly.

> > One detail I would like to discuss is whether we really want to support categories that are numerical.
> 
> Yes.

Let me clarify my question. Should we consider `3` (a number) and `'3'` (a string) to represent different categories? If not, would it be okay to stringify everything?

Reply:
> Should we consider `3` (a number) and `'3'` (a string) to represent different categories?

yes -> https://codepen.io/etpinard/pen/NVqgdZ?editors=1010

Reply:
Using `BADNUM` in ebf545ff33bd2f9fee38b516d73c891a2a2d1bee

Reply:
> > Should we consider `3` (a number) and `'3'` (a string) to represent different categories?
> 
> yes -> https://codepen.io/etpinard/pen/NVqgdZ?editors=1010

I'm confused. It seems to me that this Codepen show that  `1` and `'1'` are considered to be the **same** category.

Reply:
> It seems to me that this Codepen shows that `1` and `'1'` are considered to be the **same** category.

Yes, that's what I meant. Sorry. Edited above.